_posts/2016-03-18-parser-combinators.md

```diff
@@ -0,0 +1,470 @@
+---
+layout: post
+title: Building a lexer and parser with Scala's Parser Combinators
+---
+
+
+As part of an ongoing project at [**e.near**](http://www.enear.co/), one of our Scala teams was recently tasked with a requirement to build an interpreter for executing workflows which are modelled with a textal DSL. These workflows had to be validated for errors, compiled to a simpler bytecode-like representation and stored, in order to be interpreted with as little overhead as possible.
+
+To deal with the parsing phases, we stumbled upon [Scala's Parser Combinators](https://github.com/scala/scala-parser-combinators), which turned out to be a great tool for writing modular and composable lexers and parsers in idiomatic Scala, with little to no boilerplate code.
+
+While there are a few Web resources on how to use parser combinators for building simple parsers, there were none so far that describe how to build full lexical and syntactical analyzers from scratch.
+
+We will be going through the process of building a lexer, to scan text into a sequence of tokens, and a parser, to parse said tokens into an abstract syntax tree (AST).
+
+
+### Language overview
+
+The language we will be here is a subset of the "workflow code" we have implemented for our customer. A program implements a workflow, i.e. an executable directed acyclic graph of instructions. A workflow keeps an internal set of facts and may call services which consume facts to produce other facts.
```

Comment:
Remove the line "The language we will be here is a subset of the "workflow code" we have implemented for our customer.". The reader doesn't need to know that what we are presenting is a subset of the final language...